lib/service/connect.go

```diff
@@ -1041,20 +1047,34 @@ func (process *TeleportProcess) rotate(conn *Connector, localState auth.StateV2,
 	}
 }
 
+// getClient gets an appropriate client for the given identity. The instance client is reused if appropriate, otherwise
+// a new client is created.
+func (process *TeleportProcess) getClient(identity *auth.Identity) (clt *auth.Client, reused bool, err error) {
+	if identity.ID.Role != types.RoleInstance {
+		// non-instance roles should wait to see if the instance client can be reused
+		// before acquiring their own client.
+		if conn := process.waitForInstanceConnector(); conn != nil && conn.Client != nil {
```

Comment:
This is okay but still feels a little bit tricky. It's not immediately clear from the godoc or the signature that `reused == false` means "caller is resonsible for closing".

Since this is already a method on `*TeleportProcess`, is there any sort of way we could register a shutdown action to close the client as necessary? Then we could get rid of the extra return value and callers wouldn't have to differentiate these two cases.

Reply:
Updated the docs to clarify the meaning of `reused`.  Unfortunately, eliminating it isn't quite as simple as registering an `OnExit` operation. That'd work for most of the standard services, but the plugins in `e` use this codepath as well, and they may be created/destroyed independently from the overall process lifecycle.  We definitely could create a system for better per-system-role resource-management across persistent and ephemeral services/plugins, but it'd require some amount of replumbing of all services.  The `reused` value isn't perfect, but it's a very self-contained change.  Only `TeleportProcess.connect` and `TeleportProcess.firstTimeConnect` ever need to care about it, and only to pass it in as one of the fields in `Connector`.